### PR TITLE
[type: new feature] add Cache plugin.

### PR DESCRIPTION
**Pull request by qicz**
#3123 

Make sure that:

- [x] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor).
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] Your local test passed `mvn clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

File: shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/pom.xml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            <groupId>org.apache.shenyu</groupId>
+            <artifactId>shenyu-plugin-base</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.data</groupId>
+            <artifactId>spring-data-redis</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.lettuce</groupId>
```
Comment (yu199195) [inline shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/pom.xml:47]: this dependency sample to ratelimiter plugin， can used by common dependency？

File: shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/utils/CacheUtils.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    public static String contentTypeKey(final ServerWebExchange exchange) {
+        return dataKey(exchange) + CONTENT_TYPEKEY_SUFFIX;
+    }
+
+    /**
+     * get the cache.
+     *
+     * @return cache
+     */
+    public static ICache getCache() {
```
Comment (yu199195) [inline shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/utils/CacheUtils.java:68]: why not used SPI？

File: shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/utils/CacheUtils.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * the cache data key.
+     *
+     * @param exchange the exchange.
+     * @return data key
+     */
+    public static String dataKey(final ServerWebExchange exchange) {
+        // TODO: 2022/3/16 暂定为path路径，后面可以通过页面配置中设置key
```
Comment (yu199195) [inline shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/utils/CacheUtils.java:47]: todo mybe used by english

File: shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/redis/serializer/ShenyuSerializationContext.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+package org.apache.shenyu.plugin.cache.base.redis.serializer;
+
+import org.springframework.data.redis.serializer.RedisSerializationContext;
+import org.springframework.data.redis.serializer.RedisSerializer;
+import org.springframework.data.redis.serializer.StringRedisSerializer;
+
+/**
+ * ShenyuSerializationContext.
+ */
+public final class ShenyuSerializationContext {
```
Comment (yu199195) [inline shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/redis/serializer/ShenyuSerializationContext.java:27]: mybe rename to ShenyuRedisSerializationContext？

File: shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/pom.xml (hunk 1 above)
Comment (qicz) [reply shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/pom.xml:47]: the `shenyu-plugin-cache-base` is a common module for `cache-plugin` and `ratelimiter-plugin` now.

File: shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/utils/CacheUtils.java (hunk 2 above)
Comment (qicz) [reply shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/utils/CacheUtils.java:68]: @qurenneng @zhouyadong refactor this.

Comment (zhouyadong) [reply shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/utils/CacheUtils.java:68]: ok

Comment (qurenneng) [reply shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/utils/CacheUtils.java:68]: ok

Comment (yu199195) [reply shenyu-plugin/shenyu-plugin-cache/shenyu-plugin-cache-base/src/main/java/org/apache/shenyu/plugin/cache/base/utils/CacheUtils.java:68]: 1. ICache shuoud be used by shenyu SPI , You can refer to the current implementation，such as 'ParameterData.java'
2. CacheEnum mybe removed

**Comment by qicz** [comment]
@yu199195 pls view this pr

**Review by yu199195 — APPROVED** [review]

**Pull request merged**